archive/untar_stream.ts

```diff
@@ -0,0 +1,235 @@
+ * or cancelled.
+ *
+ * ### Understanding Compressed
+ * A tar archive may be compressed, often identified by an additional file
+ * extension, such as '.tar.gz' for gzip. This TransformStream does not support
+ * decompression which must be done before expanding the archive.
+ *
+ * @example
+ * ```ts
+ * import { UnTarStream } from "@std/archive/untar-stream";
+ *
+ * let fileWriter: WritableStreamDefaultWriter | undefined;
+ * for await (
+ *   const entry of (await Deno.open('./out.tar.gz'))
+ *     .readable
+ *     .pipeThrough(new DecompressionStream('gzip'))
+ *     .pipeThrough(new UnTarStream())
+ * ) {
+ *   if (entry.type === "header") {
+ *     fileWriter?.close();
+ *     fileWriter = (await Deno.create(entry.pathname)).writable.getWriter();
+ *   } else await fileWriter!.write(entry.data);
+ * }
+ * ```
+ */
+export class UnTarStream
+  implements TransformStream<Uint8Array, TarStreamChunk> {
+  #readable: ReadableStream<TarStreamChunk>;
+  #writable: WritableStream<Uint8Array>;
+  #gen: AsyncGenerator<Uint8Array>;
+  constructor() {
+    const { readable, writable } = new TransformStream<
+      Uint8Array,
+      Uint8Array
+    >();
+    this.#readable = ReadableStream.from(this.#untar());
+    this.#writable = writable;
+
+    this.#gen = async function* () {
+      const buffer: Uint8Array[] = [];
+      for await (
+        const chunk of readable.pipeThrough(new FixedChunkStream(512))
+      ) {
+        if (chunk.length !== 512) {
+          throw new Error("Tarball has an unexpected number of bytes");
+        }
+
+        buffer.push(chunk);
+        if (buffer.length > 2) yield buffer.shift()!;
+      }
+      if (buffer.length < 2) {
+        throw new Error("Tarball was too small to be valid");
+      }
+      if (!buffer.every((value) => value.every((x) => x === 0))) {
+        throw new Error("Tarball has invalid ending");
+      }
+    }();
+  }
+
+  async *#untar(): AsyncGenerator<TarStreamChunk> {
+    const decoder = new TextDecoder();
+    while (true) {
+      const { done, value } = await this.#gen.next();
+      if (done) break;
+
+      // Validate Checksum
+      const checksum = value.slice();
+      checksum.set(new Uint8Array(8).fill(32), 148);
+      if (
+        checksum.reduce((x, y) => x + y) !==
+          parseInt(decoder.decode(value.slice(148, 156 - 2)), 8)
+      ) throw new SyntaxError("Tarball header failed to pass checksum");
+
+      // Decode Header
+      let header: OldStyleFormat | PosixUstarFormat = {
+        name: decoder.decode(value.slice(0, 100)).replaceAll("\0", ""),
+        mode: decoder.decode(value.slice(100, 108 - 2)),
+        uid: decoder.decode(value.slice(108, 116 - 2)),
+        gid: decoder.decode(value.slice(116, 124 - 2)),
+        size: parseInt(decoder.decode(value.slice(124, 136)).trimEnd(), 8),
+        mtime: parseInt(decoder.decode(value.slice(136, 148 - 1)), 8),
+        checksum: decoder.decode(value.slice(148, 156 - 2)),
```

Comment:
Numeric fields may also be terminated by NULs besides spaces, but since `parseInt` only parses up to the first nonnumeral you don't need to remove them beforehand.

Reply:
There is no harm in the way the nulls are being removed here since the v8 engine should optimise the `x - 2` to be a new literal, and having them makes it easier, in my opinion, to calculate that they're slicing the right area.

Reply:
According to the spec,

> Each numeric field is terminated by one or more <space> or NUL characters.

Suppose the `mode` field contains `"0000644\0"`. `value.slice(100, 108 - 2)` truncates the value, while `value.slice(100, 108)` would just work because `parseInt("0000644\0", 8) === 0o644`.

Reply:
I understand that but there isn't any harm in doing `.slice(100, 108 - 2)` compared to `.slice(100, 108)`. It's just being more explicit about where the value actually is.